core/src/main/java/com/linecorp/armeria/client/ClientRequestContext.java

```diff
@@ -289,6 +295,66 @@ static ClientRequestContextBuilder builder(RpcRequest request, URI uri) {
     @Override
     RpcRequest rpcRequest();
 
+    /**
+     * Pushes the specified context to the thread-local stack. To pop the context from the stack, call
+     * {@link SafeCloseable#close()}, which can be done using a {@code try-with-resources} block:
+     * <pre>{@code
+     * try (PushHandle ignored = ctx.push(true)) {
+     *     ...
+     * }
+     * }</pre>
+     *
+     * <p>In order to call this method, the current thread-local state must meet one of the
+     * following conditions:
+     * <ul>
+     *   <li>the thread-local does not have any {@link RequestContext} in it</li>
+     *   <li>the thread-local has the same {@link ClientRequestContext} as this - reentrance</li>
+     *   <li>the thread-local has the {@link ServiceRequestContext} which is the same as {@link #root()}</li>
+     *   <li>the thread-local has the {@link ClientRequestContext} whose {@link #root()}
+     *       is the same {@link #root()}</li>
+     *   <li>the thread-local has the {@link ClientRequestContext} whose {@link #root()} is {@code null}
```

Comment:
Is this correct? They are two unrelated contexts

Reply:
I think this could happen in test cases or where the client call is not triggered by a service
```java
final WebClient client =
        WebClient.builder(server.uri("/"))
                 .decorator(RetryingHttpClient...)
                 .build();

final CompletableFuture<AggregatedHttpResponse> future = client.get("/").aggregate();
future.handle((msg, cause) -> {
    // This thread has a clientRequestContext in the thread-local. #2282
    client2.execute(...) // ctx2.push() happens while ctx1 is in the thread-local
    return null;
});
```
So I think we can allow this condition not to throw an exception. Could you have another case that you concerned about?